File: engine/openscript/accounting/analysis.py
"""The trades taken apart: by direction, by extreme, and by run.

The summary answers what the whole run did. Three questions it cannot answer are
asked here, and each of them is a question about whether the summary means what
it looks like it means.

**Which side made the money.** A run whose long trades paid for its short ones
reports a healthy net and is two strategies, one of which is losing. The summary
cannot show that, because every figure in it is folded over both sides at once.
Splitting it is not a refinement of the headline number, it is the first thing
that can contradict it.

**Whether one trade is the result.** A hundred trades and a profit factor of two
reads as an edge until the largest win is the whole of the net. The expectancy's
standard error already says how wide the spread is; the largest win and the
largest loss say where the width came from, which is the part a reader can act
on.

**What the run of losses was.** The deepest drawdown is a money figure and the
longest one is a bar count, and neither is the number that actually stops
somebody trading a strategy. That number is how many times in a row it was
wrong, and it is not derivable from anything in the summary: the same win rate
over the same trades gives a streak of two or a streak of eleven depending on an
ordering the summary folds away.

**Every figure here is over closed trades, and that is the whole rule.** An open
trade has no net to win or lose by, so it is in none of these counts, in no
streak and in no extreme. The consequence worth stating: ``long.count`` plus
``short.count`` is the summary's ``trade_count`` and not the length of the list
folded, and the summary's ``open_trade_count`` is where the difference goes.

This is the first engine's ``src/core/accounting/analysis.ts`` in this language's
spellings. The two are one algorithm and the conformance suite is what holds
them to it.
"""

from dataclasses import dataclass
from typing import List, Optional, Sequence

from .trades import Trade


@dataclass(frozen=True)
class SideAnalysis:
    """One direction's own figures, folded over that side's closed trades."""

    count: int
    wins: int
    losses: int
    #: Exactly zero net, counted in neither half, as in the summary.
    scratches: int
    #: Net after charges, which is the figure the sides are compared on.
    net_profit: float
    #: ``wins / (wins + losses)``, ``None`` where this side decided nothing.
    win_rate: Optional[float]


@dataclass(frozen=True)
class TradeAnalysis:
    """The trades by direction, by extreme and by run.

    ``long`` and ``short`` partition the closed trades, so their counts sum to
    the summary's ``trade_count`` and their nets sum to its ``net_profit``.
    """

    long: SideAnalysis
    short: SideAnalysis
    #: The best closed trade's net, zero where none closed and zero where every
    #: closed trade lost. Zero rather than ``None``, because the figure is read
    #: beside ``largest_loss`` and against the net, and a ``None`` in a column
    #: of money makes every reader handle a case that means "nothing won",
    #: which is what zero already means in this column.
    largest_win: float
    #: The worst closed trade's net as a positive magnitude, zero where none lost.
    largest_loss: float
    #: The longest run of consecutive winning closed trades, in the order they
    #: closed, and not the order they opened: that is the order the account
    #: experienced them in. The two differ whenever a trade is held across
    #: another one's whole life, which is every scaling strategy.
    max_consecutive_wins: int
    #: The same, for losses.
    max_consecutive_losses: int


@dataclass
class _SideTally:
    """A side's figures under construction, before the rates are taken."""

    count: int = 0
    wins: int = 0
    losses: int = 0
    scratches: int = 0
    net_profit: float = 0.0


def _side_of(tally: _SideTally) -> SideAnalysis:
    decided = tally.wins + tally.losses
    return SideAnalysis(
        count=tally.count,
        wins=tally.wins,
        losses=tally.losses,
        scratches=tally.scratches,
        net_profit=tally.net_profit,
        win_rate=None if decided == 0 else tally.wins / decided,
    )


def analysis_of(trades: Sequence[Trade]) -> TradeAnalysis:
    """One pass over the closed trades, in closing order.

    The list arrives in opening order, which is what the equity fold needs. The
    streaks need closing order, so the closed trades are ordered here rather
    than anywhere else: reordering the list the caller holds would change the
    equity curve.

    A trade whose net is exactly zero is a scratch, which is the summary's rule.
    A scratch **breaks** a streak without extending either one: a strategy that
    went right, flat, right was not right twice running, and counting the flat
    trade as either would make the streak a figure that depends on a rounding at
    the last digit.
    """
    long = _SideTally()
    short = _SideTally()
    largest_win = 0.0
    largest_loss = 0.0
    max_consecutive_wins = 0
    max_consecutive_losses = 0
    win_streak = 0
    loss_streak = 0

    closed: List[Trade] = [trade for trade in trades if not trade.is_open]
    # A trade closes on a bar, and two can close on the same one. The opening
    # order breaks the tie, because it is the order the list arrived in and the
    # only other fact available: a sort that is not total gives two engines two
    # different streaks from one list of trades.
    closed.sort(key=lambda trade: (
            0 if trade.closed_on_bar is None else trade.closed_on_bar,
            trade.index,
        ))

    for trade in closed:
        side = long if trade.side == "long" else short
        side.count += 1
        side.net_profit += trade.net_profit

        if trade.net_profit > 0:
            side.wins += 1
            if trade.net_profit > largest_win:
                largest_win = trade.net_profit
            win_streak += 1
            loss_streak = 0
            if win_streak > max_consecutive_wins:
                max_consecutive_wins = win_streak
        elif trade.net_profit < 0:
            side.losses += 1
            if -trade.net_profit > largest_loss:
                largest_loss = -trade.net_profit
            loss_streak += 1
            win_streak = 0
            if loss_streak > max_consecutive_losses:
                max_consecutive_losses = loss_streak
        else:
            side.scratches += 1
            win_streak = 0
            loss_streak = 0

    return TradeAnalysis(
        long=_side_of(long),
        short=_side_of(short),
        largest_win=largest_win,
        largest_loss=largest_loss,
        max_consecutive_wins=max_consecutive_wins,
        max_consecutive_losses=max_consecutive_losses,
    )

File: src/core/accounting/analysis.ts
/**
 * The trades taken apart: by direction, by extreme, and by run.
 *
 * The summary answers what the whole run did. Three questions it cannot answer
 * are asked here, and each of them is a question about whether the summary
 * means what it looks like it means.
 *
 * **Which side made the money.** A run whose long trades paid for its short
 * ones reports a healthy net and is two strategies, one of which is losing. The
 * summary cannot show that, because every figure in it is folded over both
 * sides at once. Splitting it is not a refinement of the headline number, it is
 * the first thing that can contradict it.
 *
 * **Whether one trade is the result.** A hundred trades and a profit factor of
 * two reads as an edge until the largest win is the whole of the net. The
 * expectancy's standard error already says how wide the spread is; the largest
 * win and the largest loss say where the width came from, which is the part a
 * reader can act on.
 *
 * **What the run of losses was.** The deepest drawdown is a money figure and
 * the longest one is a bar count, and neither is the number that actually stops
 * somebody trading a strategy. That number is how many times in a row it was
 * wrong. It is not derivable from anything in the summary: the same win rate
 * over the same trades gives a streak of two or a streak of eleven depending on
 * an ordering the summary folds away.
 *
 * ## Every figure here is over closed trades, and that is the whole rule
 *
 * An open trade has no net to win or lose by, so it is in none of these counts,
 * on either side, in no streak and in no extreme. `statistics.ts` has three
 * different answers to which trades a figure is counted over, because a summary
 * carries a capital figure and a bar count that are facts about the run rather
 * than about its trades. Nothing here is a fact about the run, so there is one
 * answer, and a reader can hold it.
 *
 * The consequence worth stating: `longCount + shortCount` is `tradeCount` and
 * not the length of the list folded. A run holding an open position reports
 * fewer trades here than it has, and the summary's `openTradeCount` is where
 * the difference is accounted for.
 */

import type { Money } from './shapes.js';
import type { Trade } from './trades.js';

/** One direction's own figures, folded over that side's closed trades. */
export interface SideAnalysis {
  readonly count: number;
  readonly wins: number;
  readonly losses: number;
  /** Exactly zero net, counted in neither half, as in the summary. */
  readonly scratches: number;
  /** Net after charges, which is the figure the sides are compared on. */
  readonly netProfit: Money;
  /** `wins / (wins + losses)`, null where this side decided nothing. */
  readonly winRate: number | null;
}

/**
 * The trades by direction, by extreme and by run.
 *
 * `long` and `short` partition the closed trades, so their counts sum to the
 * summary's `tradeCount` and their nets sum to its `netProfit`. That is
 * asserted rather than assumed, because a partition that stops partitioning is
 * the defect this shape exists to make visible and it would otherwise show up
 * as two panels that quietly disagree.
 */
export interface TradeAnalysis {
  readonly long: SideAnalysis;
  readonly short: SideAnalysis;
  /**
   * The best closed trade's net, zero where none closed and zero where every
   * closed trade lost.
   *
   * Zero rather than null, and it is the one place here that is worth arguing
   * about. A run whose every trade lost has no winning trade to name, and null
   * would say so. It reports zero because the figure is read beside
   * `largestLoss` and against `netProfit`, and a null in a column of money
   * forces every reader of the report to handle a case that means "nothing
   * won", which is what zero already means in this column.
   */
  readonly largestWin: Money;
  /** The worst closed trade's net as a positive magnitude, zero where none lost. */
  readonly largestLoss: Money;
  /**
   * The longest run of consecutive winning closed trades, in the order they
   * closed.
   *
   * In the order they closed, and not the order they opened, because that is
   * the order the account experienced them in and the streak is a fact about
   * what the account experienced. The two orders differ whenever a trade is
   * held across another one's whole life, which is every scaling strategy.
   */
  readonly maxConsecutiveWins: number;
  /** The same, for losses. */
  readonly maxConsecutiveLosses: number;
}

/** A side's figures under construction, before the rates are taken. */
interface SideTally {
  count: number;
  wins: number;
  losses: number;
  scratches: number;
  netProfit: Money;
}

function emptySide(): SideTally {
  return { count: 0, wins: 0, losses: 0, scratches: 0, netProfit: 0 };
}

function sideOf(tally: SideTally): SideAnalysis {
  const decided = tally.wins + tally.losses;
  return {
    count: tally.count,
    wins: tally.wins,
    losses: tally.losses,
    scratches: tally.scratches,
    netProfit: tally.netProfit,
    winRate: decided === 0 ? null : tally.wins / decided,
  };
}

/**
 * One pass over the closed trades, in closing order.
 *
 * The list arrives in opening order, which is what the equity fold needs and
 * what `tradesOf` produces. The streaks need closing order, so the closed
 * trades are ordered here rather than anywhere else: reordering the list the
 * caller holds would change the equity curve, and taking a second list in
 * opening order and calling its streaks correct would be the defect this
 * function is supposed to have caught.
 *
 * A trade whose net is exactly zero is a scratch, which is the summary's rule
 * and is applied here for the same reason. A scratch **breaks** a streak
 * without extending either one: a strategy that went right, flat, right was not
 * right twice running, and counting the flat trade as either would make the
 * streak a figure that depends on a rounding at the last digit.
 */
export function analysisOf(trades: readonly Trade[]): TradeAnalysis {
  const long = emptySide();
  const short = emptySide();
  let largestWin = 0;
  let largestLoss = 0;
  let maxConsecutiveWins = 0;
  let maxConsecutiveLosses = 0;
  let winStreak = 0;
  let lossStreak = 0;

  const closed = trades.filter((trade) => !trade.isOpen).slice();
  // A trade closes on a bar, and two can close on the same one. The opening
  // order breaks the tie, because it is the order the list arrived in and the
  // only other fact available: a sort that is not total gives two engines two
  // different streaks from one list of trades.
  closed.sort((a, b) => (a.closedOnBar ?? 0) - (b.closedOnBar ?? 0) || a.index - b.index);

  for (const trade of closed) {
    const side = trade.side === 'long' ? long : short;
    side.count += 1;
    side.netProfit += trade.netProfit;

    if (trade.netProfit > 0) {
      side.wins += 1;
      if (trade.netProfit > largestWin) largestWin = trade.netProfit;
      winStreak += 1;
      lossStreak = 0;
      if (winStreak > maxConsecutiveWins) maxConsecutiveWins = winStreak;
    } else if (trade.netProfit < 0) {
      side.losses += 1;
      if (-trade.netProfit > largestLoss) largestLoss = -trade.netProfit;
      lossStreak += 1;
      winStreak = 0;
      if (lossStreak > maxConsecutiveLosses) maxConsecutiveLosses = lossStreak;
    } else {
      side.scratches += 1;
      winStreak = 0;
      lossStreak = 0;
    }
  }

  return {
    long: sideOf(long),
    short: sideOf(short),
    largestWin,
    largestLoss,
    maxConsecutiveWins,
    maxConsecutiveLosses,
  };
}
